README.md
[![Bonsai Asset Badge](https://img.shields.io/badge/Sensu%20Go%20Load%20Check-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/Salamek/sensu-go-load-check) [![TravisCI Build Status](https://travis-ci.org/Salamek/sensu-go-load-check.svg?branch=master)](https://travis-ci.org/Salamek/sensu-go-load-check)

# Sensu Go Load Check
- [Overview](#overview)
- [Usage examples](#usage-examples)
- [Configuration](#configuration)
  - [Asset registration](#asset-registration)
  - [Asset definition](#asset-definition)
  - [Check definition](#resource-definition)
- [Installation from source and contributing](#installation-from-source-and-contributing)
- [Additional notes](#additional-notes)

## Overview

This plugin provides native load instrumentation including load health and per-core metrics. The `sensu-go-load-check` check takes the flags `-w` (warning) and `-c` (critical) and a desired L1/L5/L15. By default, these are a warning value of 2.75, 2.5, 2.0 and a critical value of 3.5, 3.25, 3.0. This check also outputs data as `nagios_perfdata`(for more information, see [this Nagios documentation article](https://assets.nagios.com/downloads/nagioscore/docs/nagioscore/3/en/perfdata.html). This allows for the check to be used as both a status check and a metric check. You can see an example of this in the [example check definition](#check-definition) below.

## Usage Examples

### Command line help

```
The Sensu Go check for system Load usage

Usage:
  sensu-go-load-check [flags]

Flags:
  -c, --critical string   Critical value for system load (default "3.5, 3.25, 3.0")
  -h, --help              help for sensu-go-load-check
  -w, --warning string    Warning value for system load. (default "2.75, 2.5, 2.0")
```

### Example Output

```bash
./sensu-go-load-check
CheckLoad OK - value = 0.14, 0.10, 0.07 | core_load_1=0.14, core_load_5=0.10, core_load_15=0.07
```

## Configuration

### Asset registration

Assets are the best way to make use of this check. If you're not using this plugin as an asset, please consider doing so! If you're using Sensu 5.13 or later, you can install this plugin as an asset by running:

`sensuctl asset add Salamek/sensu-go-load-check`

Else, you can find this asset on the [Bonsai Asset Index](https://bonsai.sensu.io/assets/Salamek/sensu-go-load-check).

### Asset definition

You can download the asset definition there, or you can do a little bit of copy/pasta and use the one below:

```yml
---
type: Asset
api_version: core/v2
metadata:
  name: sensu-go-load-check
  namespace: CHANGEME
  labels: {}
  annotations: {}
spec:
  url: https://github.com/Salamek/sensu-go-load-check/releases/download/0.0.1/sensu-go-load-check_0.0.1_linux_amd64.tar.gz
  sha512:
  filters:
  - entity.system.os == 'linux'
  - entity.system.arch == 'amd64'
```

**NOTE**: ***PLEASE ENSURE YOU UPDATE YOUR URL AND SHA512 BEFORE USING THE ASSET***. If you don't, you might just be stuck on a super old version. Don't say I didn't warn you ¯\\_(ツ)_/¯

### Check definition

Example Sensu Go definition:

**sensu-go-load-check**
```yml
type: CheckConfig
api_version: core/v2
metadata:
  name: sensu-go-load-check
  namespace: CHANGEME
spec:
  command: sensu-go-load-check
  runtime_assets:
  - sensu-go-load-check
  interval: 60
  publish: true
  output_metric_format: nagios_perfdata
  output_metric_handlers:
  - influxdb
  handlers:
  - slack
  subscriptions:
  - system
```

## Installation from source and contributing

While it's generally recommended to use an asset, you can download a copy of the handler plugin from [releases][1],
or create an executable script from this source.

From the local path of the sensu-go-load-check repository:

**sensu-go-load-check**
```
go build -o /usr/local/bin/sensu-go-load-check main.go
```
To contribute to this repo, see https://github.com/sensu/sensu-go/blob/master/CONTRIBUTING.md.

## Additional notes

### Supported operating systems

This project uses `gopsutil`, and is thus largely dependent on the systems that it supports. For this plugin, the following operating systems are supported:

* Linux
* FreeBSD
* OpenBSD
* Mac OS X
* Windows
* Solaris

[1]: https://github.com/Salamek/sensu-go-load-check/releases


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).


main.go
package main

//Import the packages we need
import (
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"

	"github.com/sensu/sensu-go/types"
	"github.com/shirou/gopsutil/cpu"
	"github.com/shirou/gopsutil/load"
	"github.com/spf13/cobra"
)

//Set up some variables. Most notably, warning and critical as load 1/5/15 values
var (
	warning, critical string
	stdin             *os.File
)

//Start our main function
func main() {
	rootCmd := configureRootCommand()
	if err := rootCmd.Execute(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

//Set up our flags for the command. Note that we have load defaults for warning & critical
func configureRootCommand() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "sensu-go-load-check",
		Short: "The Sensu Go check for system Load",
		RunE:  run,
	}

	cmd.Flags().StringVarP(&warning,
		"warning",
		"w",
		"2.75, 2.5, 2.0",
		"Load WARNING threshold, 1/5/15 min average")

	cmd.Flags().StringVarP(&critical,
		"critical",
		"c",
		"3.5, 3.25, 3.0",
		"Load CRITICAL threshold, 1/5/15 min average")

	return cmd
}

func parseArg(arg string) []float64 {
	var list []float64
	for _, x := range strings.Split(arg, ",") {
		w, err := strconv.ParseFloat(strings.TrimSpace(x), 64)
		if err != nil {
			msg := fmt.Sprintf("Failed to parse %s", err.Error())
			io.WriteString(os.Stdout, msg)
			os.Exit(3)
		}
		list = append(list, w)
	}

	return list
}

func run(cmd *cobra.Command, args []string) error {

	if len(args) != 0 {
		_ = cmd.Help()
		return fmt.Errorf("invalid argument(s) received")
	}

	if stdin == nil {
		stdin = os.Stdin
	}

	event := &types.Event{}

	return checkLoad(event)
}

//Here we start the meat of what we do.
func checkLoad(event *types.Event) error {

	const checkName = "CheckLoad"
	const metricName = "core_load"

	warn := parseArg(warning)
	crit := parseArg(critical)

	cpuLoad, err := load.Avg()
	if err != nil {
		msg := fmt.Sprintf("Failed to obtain load info %s", err.Error())
		io.WriteString(os.Stdout, msg)
		os.Exit(3)
	}

	cpuCount, err := cpu.Counts(false)
	if err != nil {
		msg := fmt.Sprintf("Failed to obtain CPU counts %s", err.Error())
		io.WriteString(os.Stdout, msg)
		os.Exit(3)
	}

	cpuLoadList := []float64{cpuLoad.Load1, cpuLoad.Load5, cpuLoad.Load15}

	// Calculate load per core
	for i, x := range cpuLoadList {
		cpuLoadList[i] = x / float64(cpuCount)
	}

	// Detect total level
	// 0=ok, 1=warn, 2=crit
	var level int = 0
	for i, v := range cpuLoadList {
		if v > crit[i] {
			if 2 > level {
				level = 2
			}
		} else if v >= warn[i] && v <= crit[i] {
			if 1 > level {
				level = 1
			}
		}
	}

	levelStrings := []string{"OK", "WARNING", "CRITICAL"}

	msg := fmt.Sprintf(
		"%[1]s %[2]s - value = %.2[4]f, %.2[5]f, %.2[6]f | %[3]s_1=%.2[4]f, %[3]s_5=%.2[5]f, %[3]s_15=%.2[6]f\n",
		checkName,
		levelStrings[level],
		metricName,
		cpuLoadList[0],
		cpuLoadList[1],
		cpuLoadList[2])
	io.WriteString(os.Stdout, msg)
	os.Exit(level)

	return nil
}
